ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1007 @@
+        logprobs = logits_to_logprobs(logits, output_sequence)
+    else:
+        logprobs = torch.tensor([])
+
+    return out_text, logprobs
+
+
+def logits_to_logprobs(logits: torch.Tensor, token_ids: torch.Tensor) -> torch.Tensor:
+    """Convert logits to log probabilities.
+
+    Roughly following https://huggingface.co/Pwicke/logprobs_for_CausalLMs
+
+    Note that PretrainedModel.generate returns a tuple of logits, but
+    TransformerHandler._process_outputs converts it into tensor for us.
+    """
+    logprobs = nn.functional.log_softmax(logits, dim=-1)
+    # Can also be written as logprobs[torch.arange(len(token_ids)), token_ids]
+    # but gather is more efficient.
+    return torch.gather(logprobs, dim=-1, index=token_ids.unsqueeze(-1)).squeeze(-1)
+
+
+def _move_tensor(x, device: torch.device) -> torch.Tensor:
+    return x.to(device) if isinstance(x, torch.Tensor) else x
+
+
+def maybe_set_tokenizer_chat_template(tokenizer: PreTrainedTokenizer) -> None:
+    """Set the chat template for the tokenizer if needed."""
+    # Check if the tokenizer's existing chat_template contains the '{% generation %}' tag
+    if tokenizer.chat_template and "{% generation %}" in tokenizer.chat_template:
+        # The tokenizer's chat_template already has the required tag
+        return
+    model_name = tokenizer.name_or_path.lower()
+    match model_name:
+        case name if "meta-llama/llama-3" in name:
+            template_path = (
+                f"{REPO_ROOT}/ldp/nn/chat_templates/llama3_chat_template.jinja"
+            )
+        case name if "meta-llama/llama-2" in name:
+            template_path = (
+                f"{REPO_ROOT}/ldp/nn/chat_templates/llama2_chat_template.jinja"
+            )
+        case _:
+            if "PYTEST_CURRENT_TEST" in os.environ and "gpt2" in model_name:
+                template_path = (
+                    f"{REPO_ROOT}/ldp/nn/chat_templates/llama3_chat_template.jinja"
+                )
+                logger.info("Using llama3 chat template for gpt2 model in tests.")
+                with open(template_path) as file:  # noqa: FURB101
+                    loaded_chat_template = file.read()
+                tokenizer.chat_template = loaded_chat_template
+
+                loaded_chat_template = loaded_chat_template.replace(
+                    "<|eot_id|>", tokenizer.eos_token
+                )
+                tokenizer.chat_template = loaded_chat_template
+                return
+
+            # Warn the user about potential training issues
+            logger.warning(
+                "Tokenizer's chat_template does not contain '{% generation %}'. "
+                "Training will have issues as the tag does not exist, which means "
+                "HuggingFace's internal code for retrieving the assistant_mask for training will be invalid."
+            )
+            return
+
+    with open(template_path) as file:  # noqa: FURB101
+        loaded_chat_template = file.read()
```

Comment:
Jw, why not listen to `FURB101`? I like this rule because it's less code

Reply:
Hm I don't remember. Switching to `template_path.open()`.